tests/test_h1_server.c

```diff
@@ -0,0 +1,409 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/http/connection.h>
+#include <aws/http/private/connection_impl.h>
+#include <aws/http/private/request_response_impl.h>
+#include <aws/http/request_response.h>
+#include <aws/http/server.h>
+
+#include <aws/common/clock.h>
+#include <aws/common/condition_variable.h>
+#include <aws/common/log_writer.h>
+#include <aws/common/uuid.h>
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/event_loop.h>
+#include <aws/io/logging.h>
+#include <aws/io/socket.h>
+#include <aws/io/tls_channel_handler.h>
+#include <aws/testing/aws_test_harness.h>
+
+#include <aws/testing/io_testing_channel.h>
+
+#if _MSC_VER
+#    pragma warning(disable : 4204) /* non-constant aggregate initializer */
+#endif
+
+#define TEST_CASE(NAME)                                                                                                \
+    AWS_TEST_CASE(NAME, s_test_##NAME);                                                                                \
+    static int s_test_##NAME(struct aws_allocator *allocator, void *ctx)
+
+/* get the request from the channel */
+struct tester_request {
+    struct aws_http_stream *request_handler;
+
+    /* All cursors in tester_request point into here */
+    struct aws_byte_buf storage;
+
+    struct aws_byte_cursor method;
+    struct aws_byte_cursor uri;
+    struct aws_http_header headers[100];
+    size_t num_headers;
+
+    bool header_done;
+    bool has_incoming_body;
+
+    bool stop_auto_window_update;
+
+    struct aws_byte_cursor body;
+};
+
+/* Singleton used by tests in this file */
+struct tester {
+    struct aws_allocator *alloc;
+    struct aws_logger logger;
+
+    struct aws_http_connection *server_connection;
+    struct testing_channel testing_channel;
+
+    struct tester_request requests[100];
+    int request_num;
+
+    bool server_connection_is_shutdown;
+
+} s_tester;
+
+static void s_test_on_complete(struct aws_http_stream *stream, int error_code, void *user_data) {
+    struct tester_request *request = user_data;
+    (void)stream;
+    struct aws_http_stream *r_handler = request->request_handler;
+    struct aws_byte_buf storage_buf = request->storage;
+
+    request->method = r_handler->incoming_request_method_str;
+    request->method.ptr = storage_buf.buffer + storage_buf.len;
```

Comment:
Please move this line after the `write` call below, it's kind of confusing. Technically this doesn't point at valid memory until after the write completes.